varstored-watch/varstored-watch.c

```diff
@@ -0,0 +1,178 @@
+                    "--uid", "416",
+                    NULL);
+    }
+    // Free the strings.
+    free(process_name);
+    free(vpid);
+    free(uuid_param);
+
+    return pid;
+}
+
+#define XS_VARSTORED_WATCH_PID_PATH "/local/domain/%s/varstored-watch-pid"
+
+static bool
+xs_write_pid(struct xs_handle *xsh, char * domid)
+{
+    char *varstore_watch_pid = NULL;
+    char *key = NULL;
+    bool ret = false;
+
+    /* write out the pid so libxl can tear us down on guest shutdown */
+    if (asprintf(&varstore_watch_pid, "%u", getpid()) != -1)
+        if (asprintf(&key, XS_VARSTORED_WATCH_PID_PATH, domid) != -1)
+            ret = xs_write(xsh, 0, key, varstore_watch_pid, strlen(varstore_watch_pid));
+
+    free(key);
+    free(varstore_watch_pid);
+    return ret;
+}
+
+static void
+usage(const char *prog)
+{
+    int i;
+
+    fprintf(stderr, "Usage: %s <options>\n\n", prog);
+
+    for (i = 0; i < 2; i++) {
+        if (varstored_watch_option[i].has_arg) {
+            fprintf(stderr, "\t--%s <val>\n",
+                    varstored_watch_option[i].name);
+        } else {
+            fprintf(stderr, "\t--%s\n", varstored_watch_option[i].name);
+        }
+    }
+
+    fprintf(stderr, "\n");
+}
+
+int
+main(int argc, char **argv)
+{
+    int rc;
+    int ret = 0;
+    struct xs_handle *xsh = NULL;
+    char *domid_str = NULL;
+    char *uuid = NULL;
+    const char *prog;
+
+    prog = basename(argv[0]);
+
+    while (1) {
+        char c;
+
+        c = getopt_long(argc, argv, "", varstored_watch_option, NULL);
+        if (c == -1)
+            break;
+
+        switch (c) {
+           case 'd':
+                domid_str = strdup(optarg);
+                break;
+            case 'u':
+                uuid = strdup(optarg);
+                break;
+            default:
+                usage(prog);
+                exit(2);
+        }
+    }
+
+    if (domid_str == NULL) {
+        usage(prog);
+        exit(2);
+    }
+
+    if (uuid == NULL) {
+        usage(prog);
+        exit(2);
+    }
+
+    xsh = xs_open(0);
+    if (!xsh) {
+        syslog(LOG_ERR, "Couldn't open xenstore: %d, %s", errno, strerror(errno));
+        goto err;
+    }
+
+    if (!xs_write_pid(xsh, domid_str)) {
+        syslog(LOG_ERR, "Failed to write pid to xenstore: %d, %s\n", errno, strerror(errno));
+        goto err;
+    }
+
+    //Start varstored for the first time.
+    varstored_pid = start_varstored(xsh, domid_str, uuid);
+
+    //Start loop for restarting varstored if it goes down.
+    while (true) {
+       waitpid(varstored_pid, &rc, 0);
+       if (WIFEXITED(rc)) {
+           syslog(LOG_ERR, "Varstored failed to start normally and exited...not retrying. Exit status %d\n", WEXITSTATUS(rc));
+           goto done;
+       } else if (WIFSIGNALED(rc)) {
+           syslog(LOG_ERR, "Varstored killed by signal %d, restarting.\n", WTERMSIG(rc));
+           varstored_pid = start_varstored(xsh, domid_str, uuid);
+       } else if (WIFSTOPPED(rc)) {
```

Comment:
```
    WIFSTOPPED(wstatus)
              returns true if the child process was stopped by delivery  of  a
              signal;  this  is  possible only if the call was done using WUN‐
              TRACED or when the child is being traced (see ptrace(2)).
```
The process hasn't died - it has just been stopped.  SIGCONT would resume it, I think?  Therefore starting a new varstored doesn't seem right.  WUNTRACED is not set in waitpid options through, so it would only happen under ptrace.  In case someone does ptrace varstored, I'd just remove the WIFSTOPPED case.

Reply:
Yeah good point. Didn't catch the nuance there, removing.